File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int r = grid.length;
        
        if (r == 0) return 0;
        
        int c = grid[0].length;
        int n = 0;
        
        for(int i = 0; i < r; i++) {
            for(int j = 0; j < c; j++) {
                if (grid[i][j] != '1') { continue; }
                helper(i, j, grid);
                n++;
            }
        }
        
        return n;
    }
    
    void helper(int r, int c, char[][] grid) {
        if (r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] != '1') {
            return;
        }
        
        grid[r][c] = '0';
        
        for(int[] n : new int[][] { {1, 0}, {0, 1}, {-1, 0}, {0, -1} }) {
            helper(r + n[0], c + n[1], grid);
        }
    }
}

File: 200-number-of-islands/200-number-of-islands.scala
object Solution {
    def numIslands(grid: Array[Array[Char]]): Int = {
        var res = 0
        
        for(r <- 0 to grid.length - 1) {
            for(c <- 0 to grid(0).length - 1) {
                if (grid(r)(c) != '0') {
                    helper(r, c, grid)
                    res += 1
                }
            }
        }
        
        res
    }
    
    def helper(r: Int, c: Int, grid: Array[Array[Char]]): Unit = {
        grid(r)(c) = '0'
        
        for(n <- Seq((1, 0), (-1, 0), (0, 1), (0, -1))) {
            val nr = r + n._1
            val nc = c + n._2
            if (nr >= 0 && nc >= 0 && nr < grid.length && nc < grid(0).length && grid(nr)(nc) != '0') {
                helper(nr, nc, grid)
            }
        }
    }
}